ursa_core/src/error.rs

```diff
@@ -0,0 +1,55 @@
+use string_error::*;
+use thiserror::Error as ThisError;
+
+/// The common errors that can occur in Ursa
+#[derive(ThisError, Debug)]
+pub enum UrsaError {
+    /// Convert IO errors
+    #[error("io error")]
+    IoError(#[from] core2::io::Error),
+    /// Convert format errors
+    #[error("fmt error")]
+    FmtError(#[from] std::fmt::Error),
+    /// Convert parse bool from string errors
+    #[error("parse bool error")]
+    ParseBoolError(#[from] std::str::ParseBoolError),
+    #[error("environment variable error")]
+    /// Convert environment variable errors
+    EnvVarError(#[from] std::env::VarError),
+    /// Convert json errors
+    #[error("json error")]
+    JsonError(#[from] serde_json::Error),
+    /// Convert cbor errors
+    #[error("cbor error")]
+    CborError(#[from] serde_cbor::Error),
+    /// Convert bare errors
+    #[error("bare error")]
+    BareError(#[from] serde_bare::error::Error),
+    /// Convert thread access errors
+    #[error("thread access error")]
+    ThreadAccessError(#[from] std::thread::AccessError),
+    /// Generic errors to handle anything that implements the std::error::Error trait
+    #[error("error: {0}")]
+    Kind(Box<dyn std::error::Error>),
+}
+
+impl From<String> for UrsaError {
+    fn from(value: String) -> Self {
+        Self::Kind(into_err(value))
+    }
+}
+
+impl From<&String> for UrsaError {
+    fn from(value: &String) -> Self {
+        Self::Kind(into_err(value.clone()))
+    }
+}
+
+impl From<&str> for UrsaError {
+    fn from(value: &str) -> Self {
+        Self::Kind(new_err(value))
+    }
+}
```

Comment:
This is odd. Usually one defines the impls in terms of `AsRef<str>`. Why have you decided to implement the conversion from String, instead of directly using &str?

Reply:
These are convenience methods to convert from a String. We don't want there to be a `core` update each time a new error is needed, thus the `Kind` value which can be anything that is derived from std::error::Error. The most common needed value is `String`.

Reply:
In that case, it might be worth considering `impl FromStr for Error`. This would also allow you to retain some strong typing via runtime checks. 

Additionally, these methods have a few issues. 

1. It's more common to have `impl From<AsRef<str>>`, which would allow you to instantiate it for more types that act like strings. 

2. `From<&String>` is allocating implicitly. Moreover, it would be covered by the `AsRef` impl, so you don't lose much by having the standard implementation. 

3. These functions are all repeated with no good reason. If I were to keep them, I'd perhaps delegate to the `&str` implementation. DRY.

Reply:
To dry this out would be better for a macro. I've had unintended consequences happen when I've done `from<impl Trait> for ...`.